kadm5: pre/post-commit plugin hook for kadm5 update operations (#397)

lhoward opened the pull request:
This change adds plugin support to the kadmin libraries for performing actions before and after a password change is committed to the KDC database and after a change is made to the attributes of a principal (specifically, a change to DISALLOW_ALL_TIX).

This change adds a hook_libraries configuration option to the [kadmin] section of krb5.conf (or kdc.conf if you use that file) that must be set to load the module. That configuration option is in the form:

```
[kadmin]
  hook_libraries = /usr/local/lib/krb5/plugins/kadm5_hook/krb5_sync.so
```

where the value is the full path to the plugin that you want to load. If this option is not present, kadmind will not load a plugin and the changes from the patch will be inactive. If this option is given and the plugin cannot be loaded, kadmind startup will abort with a (hopefully useful) error message in syslog.

Any plugin used with this patch must expose a public struct named kadm5_hook. That struct must contain the following:

```
typedef struct kadm5_hook {
    const char *name;
    int version;
    const char *vendor;

    krb5_error_code (*init)(krb5_context, void **);
    void (*fini)(krb5_context, void *);

    krb5_error_code (*chpass)(krb5_context, void *, enum kadm5_hook_stage,
                              krb5_principal, const char *);
    krb5_error_code (*create)(krb5_context, void *, enum kadm5_hook_stage,
                              kadm5_principal_ent_t, uint32_t mask,
                              const char *password);
    krb5_error_code (*modify)(krb5_context, void *, enum kadm5_hook_stage,
                              kadm5_principal_ent_t, uint32_t mask);
    ...
} kadm5_hook;
```

where enum kadm5_hook_stage is:

```
enum kadm5_hook_stage {
    KADM5_HOOK_STAGE_PRECOMMIT,
    KADM5_HOOK_STAGE_POSTCOMMIT
};
```

init creates a hook context that is passed into all subsequent calls. chpass is called for password changes, create is called for principal creation (with the newly-created principal in the kadm5_principal_ent_t argument), and modify is called when a principal is modified.

These functions should follow the normal Kerberos calling convention of returning 0 on success and a Kerberos error code on failure, setting the Kerberos error message in the provided context.

This change is submitted under the following license

Copyright 2012, 2013
The Board of Trustees of the Leland Stanford Junior University

Copying and distribution of this file, with or without modification, are permitted in any medium without royalty provided the copyright notice and this notice are preserved. This file is offered as-is, without any warranty.

jaltman commented:
`/*
 * libkadm5srv expects a symbol named kadm5_hook_v0 exported by the dynamicaly
 * loaded module and of type kadm5_hook.  version must be
 * KADM5_HOOK_VERSION_V0.  Any or all of the function pointers may be NULL, in
 * which case that hook will not be called.
 */`

I don't like the idea of exporting a data symbol as it doesn't work well on Windows. I would prefer the shared library export a function that passes the data structure as an output parameter. 

Since the plugin functions accept the krb5_context it is critical that we detect mismatches due to modifications of the krb5_context structure.   Perhaps export a function

`  int init(void **kadm_hook);`

and add a `size_t sizeof_krb5_context` parameter to the structure in addition to the kadm_hook version.

lhoward commented:
I've amended the patch to export a function instead. I'd prefer to use a version number than sizeof(krb5_context) as the hook won't have access to krb5_locl.h. If KADM5_STRUCT_VERSION changes then we can up KADM5_HOOK_VERSION.

lhoward commented:
I also have a patch for krb5-sync which compiles although I don't have the infrastructure to test it.

lhoward commented:
TOOD: use krb5_const_principal in plugin API

nicowilliams commented:
@lhoward given your TODO comment, is this ready for review?

lhoward commented:
I’ll fix it this morning.

lhoward commented:
Fixed, ready for review

nicowilliams commented on lib/kadm5/kadm5-hook.h:
```diff
@@ -0,0 +1,126 @@
+typedef struct kadm5_hook {
+    const char *name;
+    uint32_t version;
+    const char *vendor;
```
Unused.  What's it for?  Documentation?

lhoward replied:
Appears to be.

nicowilliams commented on lib/kadm5/chpass_s.c:
```diff
@@ -67,6 +68,22 @@ change(void *server_handle,
     if (ret)
 	goto out2;
 
+    for (i = 0; i < context->num_hooks; i++) {
```
The calling of the hooks could be abstracted out.

lhoward replied:
Ack.

nicowilliams commented on lib/kadm5/server_hooks.c:
```diff
@@ -0,0 +1,175 @@
+    kadm5_ret_t ret = KADM5_BAD_SERVER_HOOK;
+
+    libraries = krb5_config_get_strings(context, NULL,
+					"kadmin", "hook_libraries", NULL);
```
s/hook_libraries/hooks/

lhoward commented on lib/kadm5/kadm5-hook.h:
```diff
@@ -0,0 +1,126 @@
+    const char *name;
+    uint32_t version;
+    const char *vendor;
+    void (KRB5_CALLCONV *fini)(krb5_context, void *data);
```
I agree that passing a table of function pointers *in* is better, but as you mentioned, given we're passing `kadm5_principal_ent_t` in perhaps it's fairly tightly coupled already.

lhoward commented on lib/kadm5/kadm5-hook.h:
```diff
@@ -0,0 +1,126 @@
+					    void *data,
+					    enum kadm5_hook_stage stage,
+					    krb5_error_code code,
+					    krb5_const_principal princ,
```
I feel if we're going down that path, then the hook API should send JSON or something...

nicowilliams replied:
Someday, when I finish my PG backend, we could use JSON-serializations of HDB entries.

lhoward commented on lib/kadm5/kadm5-hook.h:
```diff
@@ -0,0 +1,126 @@
+					    krb5_error_code code,
+					    kadm5_principal_ent_t ent,
+					    uint32_t mask,
+					    const char *password);
```
Well, also for creating with keys. But perhaps that should be a separate API, I was just going on the Stanford patch.

lhoward commented on lib/kadm5/kadm5-hook.h:
```diff
@@ -0,0 +1,126 @@
+					    void *data,
+					    enum kadm5_hook_stage stage,
+					    krb5_error_code code,
+					    kadm5_principal_ent_t ent,
```
Right.

nicowilliams commented on lib/kadm5/server_hooks.c:
```diff
@@ -0,0 +1,175 @@
+
+	hook_init = dlsym(handle, "kadm5_hook_init");
+	if (hook_init == NULL) {
+	    krb5_warnx(context, "didn't find `kadm5_hook_init' symbol in `%s':"
```
I forget.  Do these go to stderr if we're running `kadmin -l`?  I fear for usability if not.

lhoward replied:
Think it depends on `[logging]`

nicowilliams replied:
OK.  We should really make `kadmin -l` force logging to also go to `stderr`, but not today.

lhoward commented on lib/kadm5/kadm5-hook.h:
```diff
@@ -0,0 +1,126 @@
+} kadm5_hook;
+
+typedef krb5_error_code
+(KRB5_CALLCONV *kadm5_hook_init_t)(krb5_context context,
```
So, we could get the *plugin* to pass us back krb5_init_context(), that we plugins can't get around it...

nicowilliams replied:
That works for me.

lhoward replied:
Is this incompatible with loading with RTLD_GROUP? Maybe we shouldn't be specifying that.

nicowilliams commented on lib/kadm5/server_hooks.c:
```diff
@@ -0,0 +1,175 @@
+
+    libraries = krb5_config_get_strings(context, NULL,
+					"kadmin", "hook_libraries", NULL);
+    if (libraries == NULL)
```
Oy, we can't tell if `krb5_config_get_string()` failed due to `ENOMEM`.  In principle we could check if `errno == ENOMEM` here.  I believe we don't check anywhere for that, so don't.

Oh, and you need to `krb5_config_free_strings(libraries)`, FYI.

nicowilliams commented on lib/kadm5/server_hooks.c:
```diff
@@ -0,0 +1,175 @@
+	free(hook_context);
+    if (handle != NULL)
+	dlclose(handle);
+
```
```
    if (libraries != NULL)
        krb5_config_free_strings(libraries);
````

nicowilliams commented:
All my comments are in.

Changes requested:

 - abstract calling the hooks somewhat
 - don't leak `libraries`
 - s/`"hook_libraries"`/`"hooks"`/
 - have the plugin output a pointer to `krb5_init_context()` so `libkadm5` can check that the plugin is using the same `libkrb5`

lhoward commented:
Fixed, also added a not-really-related commit that moves password quality checks out of `kadmindd` and `kpasswdd` and into `libkadm5`, with a view to eventually refactoring password quality checks as pre-commit hooks.

nicowilliams commented:
Well, the least we could do is retry on realm create if the random password is "weak".

jaltman commented on lib/kadm5/server_hooks.c:
```diff
@@ -0,0 +1,183 @@
+	    ret = KADM5_NEW_SERVER_HOOK_VERSION;
+	if (ret) {
+	    krb5_warnx(context, "version of loaded hook `%s' is %u"
+		       " (supported versions %u to %u)", hookpath, hook->version,
```
Should include hook vendor in the output;

jaltman commented on lib/kadm5/server_hooks.c:
```diff
@@ -0,0 +1,183 @@
+	}
+
+	if (hook->init_context != krb5_init_context) {
+	    krb5_warnx(context, "loaded hook `%s' is not linked against "
```
should include hook version and vendor in the output.

jaltman commented on lib/kadm5/server_hooks.c:
```diff
@@ -0,0 +1,183 @@
+	hook_context = NULL;
+	ctx->num_hooks++;
+    }
+
```
I think we should krb5_warnx() the hook name, version and vendor that was successfully loaded.

jaltman commented on lib/kadm5/server_hooks.c:
```diff
@@ -0,0 +1,183 @@
+kadm5_ret_t
+_kadm5_s_init_hooks(kadm5_server_context *ctx)
+{
+    return 0;
```
In the no DLOPEN case we should log a hook load failure and return an error.

nicowilliams replied:
Only if hooks are defined in krb5.conf though.

nicowilliams commented:
BTW, @lhoward, thanks for noticing and using `krb5_prepend_error_message()`.

lhoward commented:
The lack of hyphen did trouble me @jaltman :)

I'll fix the other issues today.

lhoward commented:
Additional review comments were addressed in f62b00e3

lhoward commented:
> Well, the least we could do is retry on realm create if the random password is "weak".

Something like this?

```udiff
diff --git a/kadmin/init.c b/kadmin/init.c
index 20ed93216..abf57d6cc 100644
--- a/kadmin/init.c
+++ b/kadmin/init.c
@@ -50,11 +50,12 @@ create_random_entry(krb5_principal princ,
     kadm5_ret_t ret;
     int mask = 0;
     krb5_keyblock *keys;
-    int n_keys, i;
+    int n_keys, i, n_password_tries = 1;
     char *name;
     const char *password;
     char pwbuf[512];
 
+retry_password:
     random_password(pwbuf, sizeof(pwbuf));
     password = pwbuf;
 
@@ -83,6 +84,10 @@ create_random_entry(krb5_principal princ,
     if(ret) {
        if (ret == KADM5_DUP && (flags & CRE_DUP_OK))
            goto out;
+       if (ret == KADM5_PASS_Q_DICT && n_password_tries < 5) {
+           n_password_tries++;
+           goto retry_password;
+       }
        krb5_warn(context, ret, "create_random_entry(%s): randkey failed",
                  name);
        goto out;
```

nicowilliams commented:
Yeah, something like that.  Or disable password quality checks at init time, or special-case the `K/M` principal.

lhoward commented:
Realm initialization rewrites the random passwords with real random keys, so perhaps the real solution is to create a principal with no password and no keys, before calling generating a random key.

lhoward commented:
OK, we can do this, cleaner, tests pass:

```udiff
diff --git a/kadmin/init.c b/kadmin/init.c
index 20ed93216..47201424e 100644
--- a/kadmin/init.c
+++ b/kadmin/init.c
@@ -52,11 +52,6 @@ create_random_entry(krb5_principal princ,
     krb5_keyblock *keys;
     int n_keys, i;
     char *name;
-    const char *password;
-    char pwbuf[512];
-
-    random_password(pwbuf, sizeof(pwbuf));
-    password = pwbuf;
 
     ret = krb5_unparse_name(context, princ, &name);
     if (ret) {
@@ -76,14 +71,14 @@ create_random_entry(krb5_principal princ,
        mask |= KADM5_MAX_RLIFE;
     }
     ent.attributes |= attributes | KRB5_KDB_DISALLOW_ALL_TIX;
-    mask |= KADM5_ATTRIBUTES;
+    mask |= KADM5_ATTRIBUTES | KADM5_KEY_DATA;
 
-    /* Create the entry with a random password */
-    ret = kadm5_create_principal(kadm_handle, &ent, mask, password);
+    /* Create the entry with no keys or password */
+    ret = kadm5_s_create_principal_with_key(kadm_handle, &ent, mask);
     if(ret) {
        if (ret == KADM5_DUP && (flags & CRE_DUP_OK))
            goto out;
-       krb5_warn(context, ret, "create_random_entry(%s): randkey failed",
+       krb5_warn(context, ret, "create_random_entry(%s): create failed",
                  name);
        goto out;
     }
```
